README.md
The Internet is a random, uncontrollable and sometimes even scary place... 😵
 
But fear not! 
Now you can build your own internet and decide on your own, how your backend has to answer in order
to create reproducible tests.

## Features
* Simple configuration of mockservers
* Path matching and argument parsing
* Custom builders to build tailored responses
* Supports dio and http
* Calls a recorded and can be verified
 

## Installation 
Add the following lines to `pubspec.yaml`
```yaml
dev_dependencies:
    the_internet: ^1.0.0
```

## Usage

A simple usage example:

```dart
import 'package:the_internet/the_internet.dart';

void main() {
  final internet = TheInternet();
  final server = internet.mockServer("https://demoapi.com");

  server
    ..get("/messages", body: [
      {"id": 5, "msg": "Hello"},
      {"id": 19, "msg": "World"}
    ])
    ..post(
      "/messages",
      bodyBuilder: (request) => request.body?.asJson + {"id": 17},
    )
    ..delete("/messages/{id}", code: 204);
}
```

This will create a mockserver, which will respond to "GET /messages", "POST /messages" and 
"DELETE /message/{id}". In order to use this server instead of the real one you need to use 
a special implementation of the http or dio client:
```dart
import 'package:dio/dio.dart' as dio;
import 'package:http/http.dart' as http;

void main() {
  /* ... */

  final http.BaseClient httpClient = internet.createHttpClient();
  final dio.Dio dioClient = dio.Dio()
    ..httpClientAdapter = internet.createDioAdapter();
}
```

## Examples

#### A simple POST with http client
```dart
  test("create a message with http", () async {
    // Install a handler for POST /messages
    server.post("/messages", code: 201);

    // create a new message via http client
    final uri = Uri.parse("https://demoapi.com/messages");
    final response = await httpClient.post(uri, body: {"title": "Hello"});

    // verify the mocked response
    expect(response.statusCode, 201);
    expect(response.body, isEmpty);

    // verify the request captured by the handler
    final recordedCall = server.nextCapturedCall();

    expect(recordedCall.request.uri, uri);
    expect(recordedCall.request.body?.asString, "title=Hello");
    expect(recordedCall.request.body?.asFormData?["title"], "Hello");
    expect(recordedCall.request.body?.asJson, isNull);
  });
```

### A simple GET with dio client
```dart
  test("get messages with dio", () async {
    // Install a handler for GET /messages
    server.get("/messages", body: [
      {"id": 5, "msg": "Hello"},
      {"id": 19, "msg": "World"},
    ]);

    // get messages via dio client
    final uri = Uri.parse("https://demoapi.com/messages");
    final response = await dioClient.get("https://demoapi.com/messages");

    // verify the mocked response
    expect(response.statusCode, 200);
    expect(response.data, [
      {"id": 5, "msg": "Hello"},
      {"id": 19, "msg": "World"},
    ]);

    // verify the request captured by the handler
    final recordedCall = server.nextCapturedCall();

    expect(recordedCall.request.uri, uri);
    expect(recordedCall.request.body, isNull);
  });
```

### Response based on request
```dart
  test("build a complex response", () async {
    // Install a handler for GET /messages/{id}
    server.get("/messages/{id}", responseBuilder: (request) {
      final id = int.tryParse(request.args["id"] ?? "");
      if (id != null) {
        return MockedResponse.fromJson({
          "id": id,
          "title": "Message $id",
        });
      } else {
        return MockedResponse(404);
      }
    });

    // get a bad message via http client
    final badResponse = await httpClient.get(
      Uri.parse("https://demoapi.com/messages/not-a-number"),
    );

    // verify the mocked response
    expect(badResponse.statusCode, 404);
    expect(badResponse.body, isEmpty);

    // get a correct message via http client
    final messageResponse = await httpClient.get(
      Uri.parse("https://demoapi.com/messages/17"),
    );

    // verify the mocked response
    expect(messageResponse.statusCode, 200);
    expect(messageResponse.body, '{"id":17,"title":"Message 17"}');
  });
```
### Queuing handlers
```dart
  test("handlers can be queued", () async {
    // Install a handler for a single GET /ping
    server.get("/ping", code: 204, times: 1);
    // Install a second handler for any subsequent GET /ping
    server.get("/ping", code: 404);

    // call /ping a couple of times
    final uri = Uri.parse("https://demoapi.com/ping");
    final response1 = await httpClient.get(uri);
    final response2 = await httpClient.get(uri);
    final response3 = await httpClient.get(uri);

    // verify the response codes
    expect(response1.statusCode, 204);
    expect(response2.statusCode, 404);
    expect(response3.statusCode, 404);
  });
```



## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/avbk/the_internet/issues


CHANGELOG.md

## 1.0.0 (2021-03-07)
 - First production ready version

## 0.0.0 (2020-03-05)
- Initial idea, not production ready

test/get_test.dart
import 'package:test/test.dart';
import 'package:the_internet/the_internet.dart';

import 'test_utils.dart';

void main() {
  multiClientTestGroup("GET", (test) {
    test(
      "uses 200 <EMPTY> as default",
      configure: (server) => server.get("/nothing"),
      request: {"path": "/nothing"},
      response: {"code": 200, "body": isEmpty, "headers": isEmpty},
    );

    test(
      "allows to create a very simple GET mock",
      configure: (server) => server.get("/messages/17", code: 404),
      request: {"path": "/messages/17"},
      response: {"code": 404, "body": isEmpty, "headers": isEmpty},
    );

    test(
      "allows to respond with string",
      configure: (server) => server.get("/messages", body: "Hello World!"),
      request: {"path": "/messages"},
      response: {"code": 200, "body": "Hello World!", "headers": isEmpty},
    );

    test(
      "allows to respond with string and custom code and headers",
      configure: (server) => server.get(
        "/messages",
        code: 417,
        body: "Hello World!",
        headers: {"X-Server-Type": "Mock"},
      ),
      request: {"path": "/messages"},
      response: {
        "code": 417,
        "body": "Hello World!",
        "headers": {"X-Server-Type": "Mock"}
      },
    );

    test(
      "allows to respond with json",
      configure: (server) => server.get(
        "/messages",
        body: {
          "page": 1,
          "messages": ["Hello", "World"],
        },
      ),
      request: {"path": "/messages"},
      response: {
        "code": 200,
        "body": '{"page":1,"messages":["Hello","World"]}',
        "headers": {"Content-Type": "application/json"}
      },
    );

    test(
      "allows to respond with json and custom code and headers",
      configure: (server) => server.get("/messages", code: 417, headers: {
        "X-Server-Type": "Mock"
      }, body: {
        "page": 1,
        "messages": ["Hello", "World"],
      }),
      request: {"path": "/messages"},
      response: {
        "code": 417,
        "body": '{"page":1,"messages":["Hello","World"]}',
        "headers": {"Content-Type": "application/json", "X-Server-Type": "Mock"}
      },
    );

    test(
      "allows to respond with json based on regex",
      configure: (server) => server.get("/messages/{name}",
          bodyBuilder: (request) =>
              {"message": "Hello ${request.args["name"]}"}),
      request: {"path": "/messages/people"},
      response: {
        "code": 200,
        "body": '{"message":"Hello people"}',
        "headers": {"Content-Type": "application/json"}
      },
    );

    test(
      "allows to respond with json based on regex and headers",
      configure: (server) => server.get("/messages/{name}",
          bodyBuilder: (request) => {
                "message":
                    "${request.headers["greeting"]} ${request.args["name"]}"
              }),
      request: {
        "path": "/messages/people",
        "headers": {"greeting": "Hi"},
      },
      response: {
        "code": 200,
        "body": '{"message":"Hi people"}',
        "headers": {"Content-Type": "application/json"}
      },
    );

    test(
      "allows to respond with anything based on regex and headers",
      configure: (server) {
        server.get(
          "/messages/{name}",
          responseBuilder: (request) => MockedResponse(200,
              body: "${request.headers["greeting"]} ${request.args["name"]}"),
        );
      },
      request: {
        "path": "/messages/people",
        "headers": {"greeting": "Hi"},
      },
      response: {
        "code": 200,
        "body": "Hi people",
        "headers": isEmpty,
      },
    );

    test(
      "Query Parameters are supported",
      configure: (server) {
        server.get("/messages{?filter}",
            bodyBuilder: (request) => request.args["filter"] ?? "");
      },
      request: {
        "path": "/messages",
        "query": {"filter": "all"},
      },
      response: {
        "code": 200,
        "body": "all",
        "headers": isEmpty,
      },
    );
  });
}


test/put_test.dart
import 'package:test/test.dart';

import 'test_utils.dart';

void main() {
  multiClientTestGroup("PUT", (test) {
    test("respond with input data and check recorded call",
        configure: (server) => server.put("/messages",
            bodyBuilder: (request) => {
                  "newMessage": "message ${request.body?.asFormData?["title"]}"
                }),
        request: {
          "path": "/messages",
          "formData": {"title": "Hello"}
        },
        response: {
          "code": 200,
          "body": '{"newMessage":"message Hello"}',
          "headers": {"Content-Type": "application/json"}
        },
        recorded: {
          "request": {
            "url": "https://example.com/messages",
            "bodyAsString": "title=Hello",
            "bodyAsFormData": {"title": "Hello"},
            "bodyAsJson": isNull
          },
          "response": {
            "code": 200,
            "body": '{"newMessage":"message Hello"}',
            "headers": {"Content-Type": "application/json"}
          },
        });
  });
}


test/head_test.dart
import 'package:test/test.dart';

import 'test_utils.dart';

void main() {
  multiClientTestGroup("HEAD", (test) {
    test(
      "respond with 204",
      configure: (server) => server.head(
        "/messages/17",
        code: 204,
        headers: {"X-Server-Type": "Mocked"},
      ),
      request: {"path": "/messages/17"},
      response: {
        "code": 204,
        "body": isEmpty,
        "headers": {"X-Server-Type": "Mocked"},
      },
    );
  });
}


lib/src/call_handler.dart
part of "../the_internet.dart";

class _CallHandler {
  final String method;
  final UriParser _pathParser;
  final ResponseBuilder _buildResponse;
  final Duration? delay;
  int? times;

  _CallHandler(
    this.method,
    String pathTemplate,
    this._buildResponse,
    this.times,
    this.delay,
  ) : this._pathParser = UriParser(UriTemplate("$pathTemplate"));

  Future<MockedResponse?> _tryHandle(CapturedRequest request) async {
    if (_canTick) {
      if (request.method.toUpperCase() == method.toUpperCase()) {
        if (_pathParser.matches(request.uri)) {
          try {
            request.args = _pathParser.parse(request.uri);

            _tick();
            await _throttle();

            return _buildResponse(request);
          } on ParseException catch (_) {}
        }
      }
    }

    return null;
  }

  bool get _canTick => (times ?? 1) > 0;

  void _tick() {
    final times = this.times;
    if (times != null) {
      this.times = times - 1;
    }
  }

  Future<void> _throttle() async {
    final delay = this.delay;
    if (delay != null) {
      await Future.delayed(delay);
    }
  }
}


lib/src/the_internet.dart
part of "../the_internet.dart";

const _kDefaultCode = 200;
const _kDefaultBody = "";
const _kDefaultHeaders = <String, String>{};

/// [TheInternet] replaces the real internet with a local mocked and controlled
/// version.
///
/// Use [mockServer] to create a [MockedServer], which then can be configured:
/// ```dart
///   TheInternet internet = TheInternet();
///   MockedServer backend = internet.mockServer("https://backend.com/api");
///   backend.get("/task/{id}", code: 404, body: {"error: "Unknown Task"});
/// ```
///
/// [TheInternet] supports mocking the real internet for [http] as well as
/// for [dio]
///   * Use [createHttpClient] for [http] to create a [BaseClient] that only
///   communicates with [TheInternet].
///   * Use [createDioAdapter] for [dio] to create a [HttpClientAdapter] that
///   only communicates with [TheInternet]
///
/// ```dart
/// BaseClient client = internet.createHttpClient();
/// Response response = await client.get("/task/14");
///
/// expect(response.statusCode, 404);
/// ```
class TheInternet {
  final Map<String, MockedServer> _servers;

  /// Creates an instance of [TheInternet]
  ///
  /// Usually only one instance is needed, but it is possible to have multiple
  /// [TheInternet]s at the same time. Note that dio/http clients created in one
  /// won't be able to access [MockedServer]s of other [TheInternet]
  /// instances.
  TheInternet() : _servers = {};

  /// Creates a mocked version of the [BaseClient] that only
  /// communicates with [TheInternet].
  http.BaseClient createHttpClient() => MockClient(_handleHttpRequest);

  /// Creates a mocked version of the [HttpClientAdapter] that only
  /// communicates with [TheInternet].
  ///
  /// This adapter needs to be assigned to the [HttpClientAdapter] Property of
  /// the [Dio] instance.
  dio.HttpClientAdapter createDioAdapter() => _MockDioAdapter(this);

  /// Returns a [MockedServer] for a given baseUrl.
  ///
  /// If there already exists a server fo the given baseUrl, the
  /// existing [MockedServer] is returned.
  MockedServer mockServer(String baseUrl) {
    if (!_servers.containsKey(baseUrl)) {
      _servers[baseUrl] = MockedServer._(baseUrl);
    }
    return _servers[baseUrl]!;
  }

  /// Get a human readable summary about all the [MockedServer]s and the
  /// installed handlers.
  String get humanReadableMatchers {
    String message = "";
    for (var server in _servers.values) {
      message += "${server._host}\n";
      for (var methodAndPath in server._handlers.keys) {
        message += "\t${methodAndPath}\n";
      }
    }
    return message;
  }

  /// Resets [TheInternet].
  ///
  /// It will also reset all the registered [MockedServer]s before dropping
  /// them afterwards.
  void reset() {
    for (var server in _servers.values) {
      server.reset();
    }
    _servers.clear();
  }

  Future<http.Response> _handleHttpRequest(http.Request request) async {
    final CapturedRequest req = CapturedRequest._fromHttp(request);
    final MockedResponse response = await _handleCapturedRequest(req);
    return response._toHttp();
  }

  Future<dio.ResponseBody> _handleDioRequest(dio.RequestOptions request) async {
    final CapturedRequest req = CapturedRequest._fromDio(request);
    final MockedResponse response = await _handleCapturedRequest(req);
    return response._toDio();
  }

  Future<MockedResponse> _handleCapturedRequest(CapturedRequest request) async {
    for (var server in _servers.values) {
      final MockedResponse? response = await server._tryHandle(request);
      if (response != null) {
        return response;
      }
    }

    throw EndOfTheInternetError(this, request);
  }
}

/// This Error is thrown whenever a request is received which cannot be
/// handled by [TheInternet].
///
/// The [CapturedRequest] and [TheInternet] are provided with this error
/// for further investigation.
class EndOfTheInternetError implements Exception {
  final TheInternet internet;
  final CapturedRequest request;

  EndOfTheInternetError(this.internet, this.request);

  String get message => """
Could not find a handler for: ${request.method} ${request.uri}

Installed handlers:
${internet.humanReadableMatchers}
""";

  @override
  String toString() {
    return "EndOfTheInternetError: $message";
  }
}

class _MockDioAdapter extends dio.HttpClientAdapter {
  final TheInternet _internet;

  _MockDioAdapter(this._internet);

  @override
  Future<dio.ResponseBody> fetch(dio.RequestOptions options,
      Stream<List<int>> requestStream, Future? cancelFuture) {
    return _internet._handleDioRequest(options);
  }

  @override
  void close({bool force = false}) {}
}


lib/src/captured_call.dart
part of "../the_internet.dart";

/// A call captured by [TheInternet].
///
/// Every request is logged with its corresponding mocked response for later
/// verification.
class CapturedCall {
  /// The request, which has been received by [TheInternet] and was handled by
  /// an installed handler.
  final CapturedRequest request;

  /// The response, which has been created by an installed handler.
  final MockedResponse response;

  CapturedCall._(this.request, this.response);
}

/// A captured request.
///
/// The request, which has been captured by [TheInternet].
class CapturedRequest {
  /// The captured headers, may be empty.
  ///
  /// _Note: the fact that Dio supports  multiple values per header
  /// has not been targeted._
  final Map<String, String> headers;

  /// The captured body or null if no body was provided (e.g. a GET request)
  final CapturedBody? body;

  /// The http verb of the captured request, always uppercase
  final String method;

  /// The full uri of the captured request
  final Uri uri;

  /// All uri arguments captured while parsing the uri.
  ///
  /// The framework uses [UriParser] to match uris. If an uri matches the
  /// template of a handler, any argument in the uri is added to this map.
  ///
  /// See
  /// [pub.dev/packages/uri#uriparser](https://pub.dev/packages/uri#uriparser)
  /// for more info.
  Map<String, String> args;

  CapturedRequest._fromHttp(http.Request request)
      : this.headers = request.headers,
        this.method = request.method.toUpperCase(),
        this.uri = request.url,
        this.body = CapturedBody._fromHttp(request),
        this.args = {};

  CapturedRequest._fromDio(dio.RequestOptions request)
      : this.headers = _convertHeaders(request.headers),
        this.method = request.method.toUpperCase(),
        this.uri = request.uri,
        this.body = CapturedBody._fromDio(request),
        this.args = {};

  static Map<String, String> _convertHeaders(Map<String, dynamic> headers) =>
      headers.map((key, value) => MapEntry(
            key,
            value.toString(),
          ));
}

/// A captured body.
///
/// The body, which has been captured by [TheInternet].
class CapturedBody {
  /// The raw body as [String]
  final String asString;

  /// The data in FormData format, if the client sent form-data, otherwise null
  final Map<String, String>? asFormData;

  /// The data decoded from JSON, if the client sent JSON, otherwise null
  final dynamic asJson;

  CapturedBody._(this.asString, this.asFormData, this.asJson);

  static CapturedBody? _fromHttp(http.Request request) {
    if (request.contentLength == 0) {
      return null;
    } else {
      String body = request.body;

      Map<String, String>? formData;
      try {
        formData = request.bodyFields;
      } catch (ignored) {}

      dynamic json;
      try {
        json = jsonDecode(body);
      } catch (ignored) {}

      return CapturedBody._(body, formData, json);
    }
  }

  static CapturedBody? _fromDio(dio.RequestOptions request) {
    if (request.data is String) {
      return CapturedBody._(request.data, null, null);
    } else if (request.data is dio.FormData) {
      final formData = request.data as dio.FormData;

      return CapturedBody._(
          dio.Transformer.urlEncodeMap(Map.fromEntries(formData.fields)),
          Map.fromEntries(formData.fields),
          null);
    } else if (request.data != null) {
      return CapturedBody._(jsonEncode(request.data), null, request.data);
    } else {
      return null;
    }
  }
}

/// A captured request.
///
/// The request, which has been captured by [TheInternet].
class MockedResponse {
  /// The status code to be delivered to the client, defaults to `200`
  final int code;

  /// The body to be delivered to the client, defaults to `""`
  final String body;

  /// The headers to be delivered to the client, defaults to `{}`
  final Map<String, String> headers;

  /// Constructs an arbitrary mocked response.
  MockedResponse(
    int code, {
    String body = _kDefaultBody,
    Map<String, String> headers = _kDefaultHeaders,
  })  : this.code = code,
        this.headers = headers,
        this.body = body;

  /// Constructs a mocked JSON response.
  ///
  /// This means that the body is encoded as json and the headers are
  /// extended by `Content-Type: application/json`
  MockedResponse.fromJson(
    dynamic json, {
    int code = _kDefaultCode,
    Map<String, String> headers = _kDefaultHeaders,
  }) : this(
          code,
          body: jsonEncode(json),
          headers: Map.of(headers)
            ..addAll({"Content-Type": "application/json"}),
        );

  http.Response _toHttp() => http.Response(body, code, headers: headers);

  dio.ResponseBody _toDio() =>
      dio.ResponseBody.fromString(body, code, headers: _toDioHeaders(headers));

  Map<String, List<String>> _toDioHeaders(Map<String, String> headers) =>
      headers.map((key, value) => MapEntry(
            key.toLowerCase(),
            [value],
          ));
}


lib/src/the_internet_base.dart
import 'dart:convert';

import 'package:dio/dio.dart' as dio;
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';

const _kDefaultCode = 200;
const _kDefaultBody = "";
const _kDefaultHeaders = <String, String>{};

class TheInternet {
  final Map<String, MockedServer> _servers;

  TheInternet() : _servers = {};

  http.BaseClient createHttpClient() => MockClient(_handleHttpRequest);

  dio.HttpClientAdapter createDioAdapter() => _MockDioAdapter(this);

  MockedServer mockServer(String baseUrl) {
    _servers[baseUrl] = MockedServer._(baseUrl);
    return _servers[baseUrl];
  }

  Future<http.Response> _handleHttpRequest(http.Request request) async {
    final CapturedRequest req = CapturedRequest.fromHttp(request);
    final MockedResponse response = await _handleCapturedRequest(req);
    return response?.toHttp();
  }

  Future<dio.ResponseBody> _handleDioRequest(dio.RequestOptions request) async {
    final CapturedRequest req = CapturedRequest.fromDio(request);
    final MockedResponse response = await _handleCapturedRequest(req);
    return response?.toDio();
  }

  Future<MockedResponse> _handleCapturedRequest(CapturedRequest request) async {
    for (var server in _servers.values) {
      final MockedResponse response = server._tryHandle(request);
      if (response != null) {
        return response;
      }
    }

    return null;
  }
}

class _MockDioAdapter extends dio.HttpClientAdapter {
  final TheInternet _internet;

  _MockDioAdapter(this._internet);

  @override
  Future<dio.ResponseBody> fetch(dio.RequestOptions options,
      Stream<List<int>> requestStream, Future cancelFuture) {
    return _internet._handleDioRequest(options);
  }

  @override
  void close({bool force = false}) {}
}

typedef ResponseBuilder = MockedResponse Function(
  CapturedRequest request,
  List<String> args,
);
typedef SimpleJsonResponseBuilder = dynamic Function(List<String> args);
typedef ComplexJsonResponseBuilder = dynamic Function(
  CapturedRequest request,
  List<String> args,
);

class MockedServer {
  final String _baseUrl;
  final Map<String, _CallHandler> _handlers;
  final List<CapturedCall> _callQueue;

  MockedServer._(this._baseUrl)
      : _handlers = {},
        _callQueue = [];

  void post(
    String pathRegex, {
    int code: _kDefaultCode,
    Map<String, String> headers: _kDefaultHeaders,
    dynamic body,
    ResponseBuilder response,
  }) {
    ResponseBuilder builder = _chooseBuilder(response, body, code, headers);

    _handlers["POST $pathRegex"] =
        _CallHandler("POST", _baseUrl, pathRegex, builder);
  }

  void get(
    String pathRegex, {
    int code: _kDefaultCode,
    Map<String, String> headers: _kDefaultHeaders,
    dynamic body,
    ResponseBuilder response,
  }) {
    ResponseBuilder builder = _chooseBuilder(response, body, code, headers);

    _handlers["GET $pathRegex"] =
        _CallHandler("GET", _baseUrl, pathRegex, builder);
  }

  ResponseBuilder _chooseBuilder(
      ResponseBuilder response, body, int code, Map<String, String> headers) {
    ResponseBuilder builder;
    if (response != null) {
      builder = response;
    } else {
      if (body is String) {
        builder = (request, args) => MockedResponse(
              code,
              body: body,
              headers: headers,
            );
      } else if (body is SimpleJsonResponseBuilder) {
        builder = (request, args) =>
            MockedResponse.fromJson(
              body(args),
              code: code,
              headers: headers,
            );
      } else if (body is ComplexJsonResponseBuilder) {
        builder = (request, args) =>
            MockedResponse.fromJson(
              body(request, args),
              code: code,
              headers: headers,
            );
      } else if (body != null) {
        builder = (request, args) =>
            MockedResponse.fromJson(
              body,
              code: code,
              headers: headers,
            );
      } else {
        builder = (request, args) => MockedResponse(code);
      }
    }
    return builder;
  }

  MockedResponse _tryHandle(CapturedRequest request) {
    for (var handler in _handlers.values) {
      final MockedResponse response = handler._tryHandle(request);
      if (response != null) {
        _callQueue.add(CapturedCall(request, response));
        return response;
      }
    }

    return null;
  }

  CapturedCall nextCapturedCall() {
    return _callQueue.removeLast();
  }
}

class _CallHandler {
  final String method;
  final RegExp _regex;
  final ResponseBuilder _buildResponse;

  _CallHandler(this.method,
      String baseUrl,
      String pathRegex,
      this._buildResponse,) : this._regex = RegExp("$baseUrl$pathRegex");

  MockedResponse _tryHandle(CapturedRequest request) {
    if (request.method.toUpperCase() == method.toUpperCase()) {
      final match = _regex.firstMatch(request.url);
      if (match != null) {
        final args = List.generate(
          match.groupCount,
          // +1, because 0 is the complete pattern
              (index) => match.group(index + 1),
        );
        return _buildResponse(request, args);
      }
    }

    return null;
  }
}

class CapturedCall {
  final CapturedRequest request;
  final MockedResponse response;

  CapturedCall(this.request, this.response);
}

class CapturedRequest {
  final Map<String, String> headers;
  final CapturedBody body;
  final String method;
  final String url;

  CapturedRequest.fromHttp(http.Request request)
      : this.headers = request.headers,
        this.method = request.method,
        this.url = request.url.toString(),
        this.body = CapturedBody.fromHttp(request);

  CapturedRequest.fromDio(dio.RequestOptions request)
      : this.headers = _convertHeaders(request.headers),
        this.method = request.method,
        this.url = request.uri.toString(),
        this.body = CapturedBody.fromDio(request);

  static Map<String, String> _convertHeaders(Map<String, dynamic> headers) =>
      headers == null
          ? {}
          : Map.fromEntries(
        headers.entries.map((entry) =>
            MapEntry<String, String>(
              entry.key,
              entry.value.toString(),
            )),
      );
}

class CapturedBody {
  final dynamic asString;
  final Map<String, String> asFormData;
  final dynamic asJson;

  CapturedBody._(this.asString, this.asFormData, this.asJson);

  factory CapturedBody.fromHttp(http.Request request) {
    if (request.contentLength == 0) {
      return null;
    } else {
      String body = request.body;

      Map<String, String> formData;
      try {
        formData = request.bodyFields;
      } catch (ignored) {}

      dynamic json;
      try {
        json = jsonDecode(body);
      } catch (ignored) {}

      return CapturedBody._(body, formData, json);
    }
  }

  factory CapturedBody.fromDio(dio.RequestOptions request) {
    if (request.data is String) {
      return CapturedBody._(request.data, null, null);
    } else if (request.data is dio.FormData) {
      final formData = request.data as dio.FormData;

      return CapturedBody._(
          formData.fields.map((e) => "${e.key}=${e.value}").join("&"),
          Map.fromEntries(formData.fields),
          null);
    } else if (request.data != null) {
      return CapturedBody._(jsonEncode(request.data), null, request.data);
    } else
      return null;
  }
}

class MockedResponse {
  final int code;
  final String body;
  final Map<String, String> headers;

  MockedResponse(this.code, {
    this.body = _kDefaultBody,
    this.headers = _kDefaultHeaders,
  });

  MockedResponse.fromJson(dynamic json, {
    int code: _kDefaultCode,
    Map<String, String> headers: _kDefaultHeaders,
  }) : this(
    code,
    body: jsonEncode(json),
    headers: Map.of(headers)
      ..addAll({"Content-Type": "application/json"}),
  );

  http.Response toHttp() => http.Response(body, code, headers: headers);

  dio.ResponseBody toDio() =>
      dio.ResponseBody.fromString(body, code, headers: _toDio(headers));

  Map<String, List<String>> _toDio(Map<String, String> headers) =>
      headers ==
          null
          ? null
          : Map.fromEntries(
          headers.entries.map((e) => MapEntry(e.key.toLowerCase(), [e.value])));
}


example/the_internet_example.dart
import 'package:dio/dio.dart' as dio;
import 'package:http/http.dart' as http;
import 'package:test/test.dart';
import 'package:the_internet/the_internet.dart';

void main() {
  late TheInternet internet;
  late MockedServer server;
  late http.BaseClient httpClient;
  late dio.Dio dioClient;

  setUp(() {
    internet = TheInternet();

    // Mock a server
    server = internet.mockServer("https://demoapi.com");

    httpClient = internet.createHttpClient();
    dioClient = dio.Dio();
    dioClient.httpClientAdapter = internet.createDioAdapter();
  });

  tearDown(() {
    internet.reset();
  });

  test("create a message with http", () async {
    // Install a handler for POST /messages
    server.post("/messages", code: 201);

    // create a new message via http client
    final uri = Uri.parse("https://demoapi.com/messages");
    final response = await httpClient.post(uri, body: {"title": "Hello"});

    // verify the mocked response
    expect(response.statusCode, 201);
    expect(response.body, isEmpty);

    // verify the request captured by the handler
    final recordedCall = server.nextCapturedCall();

    expect(recordedCall.request.uri, uri);
    expect(recordedCall.request.body?.asString, "title=Hello");
    expect(recordedCall.request.body?.asFormData?["title"], "Hello");
    expect(recordedCall.request.body?.asJson, isNull);
  });

  test("get messages with dio", () async {
    // Install a handler for GET /messages
    server.get("/messages", body: [
      {"id": 5, "msg": "Hello"},
      {"id": 19, "msg": "World"},
    ]);

    // get messages via dio client
    final uri = Uri.parse("https://demoapi.com/messages");
    final response = await dioClient.get("https://demoapi.com/messages");

    // verify the mocked response
    expect(response.statusCode, 200);
    expect(response.data, [
      {"id": 5, "msg": "Hello"},
      {"id": 19, "msg": "World"},
    ]);

    // verify the request captured by the handler
    final recordedCall = server.nextCapturedCall();

    expect(recordedCall.request.uri, uri);
    expect(recordedCall.request.body, isNull);
  });

  test("build a complex response", () async {
    // Install a handler for GET /messages/{id}
    server.get("/messages/{id}", responseBuilder: (request) {
      final id = int.tryParse(request.args["id"] ?? "");
      if (id != null) {
        return MockedResponse.fromJson({
          "id": id,
          "title": "Message $id",
        });
      } else {
        return MockedResponse(404);
      }
    });

    // get a bad message via http client
    final badResponse = await httpClient.get(
      Uri.parse("https://demoapi.com/messages/not-a-number"),
    );

    // verify the mocked response
    expect(badResponse.statusCode, 404);
    expect(badResponse.body, isEmpty);

    // get a correct message via http client
    final messageResponse = await httpClient.get(
      Uri.parse("https://demoapi.com/messages/17"),
    );

    // verify the mocked response
    expect(messageResponse.statusCode, 200);
    expect(messageResponse.body, '{"id":17,"title":"Message 17"}');
  });

  test("handlers can be queued", () async {
    // Install a handler for a single GET /ping
    server.get("/ping", code: 204, times: 1);
    // Install a second handler for any subsequent GET /ping
    server.get("/ping", code: 404);

    // call /ping a couple of times
    final uri = Uri.parse("https://demoapi.com/ping");
    final response1 = await httpClient.get(uri);
    final response2 = await httpClient.get(uri);
    final response3 = await httpClient.get(uri);

    // verify the response codes
    expect(response1.statusCode, 204);
    expect(response2.statusCode, 404);
    expect(response3.statusCode, 404);
  });
}
